src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.ReadWrite.cs

```diff
@@ -0,0 +1,476 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.HttpSys.Internal;
+using System;
+using System.Buffers;
+using System.IO.Pipelines;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Server.IISIntegration
+{
+    internal partial class IISHttpContext
+    {
+        /// <summary>
+        /// Reads data from the Input pipe to the user.
+        /// </summary>
+        /// <param name="buffer"></param>
+        /// <param name="offset"></param>
+        /// <param name="count"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        public async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            // Start a task which will continuously call ReadFromIISAsync and WriteToIISAsync
+            StartProcessingRequestAndResponseBody();
+
+            while (true)
+            {
+                var result = await Input.Reader.ReadAsync();
+                var readableBuffer = result.Buffer;
+                try
+                {
+                    if (!readableBuffer.IsEmpty)
+                    {
+                        var actual = Math.Min(readableBuffer.Length, count);
+                        readableBuffer = readableBuffer.Slice(0, actual);
+                        readableBuffer.CopyTo(buffer);
+                        return (int)actual;
+                    }
+                    else if (result.IsCompleted)
+                    {
+                        return 0;
+                    }
+                }
+                finally
+                {
+                    Input.Reader.AdvanceTo(readableBuffer.End, readableBuffer.End);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Writes data to the output pipe.
+        /// </summary>
+        /// <param name="data"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        public Task WriteAsync(ArraySegment<byte> data, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (!_hasResponseStarted)
+            {
+                return WriteAsyncAwaited(data, cancellationToken);
```

Comment:
Was it always this way? we take the slow path if the response hasn't started?

Reply:
Yeah it was always this way. Suggestions to fix in the future?